java/arcs/core/analysis/Deduction.kt

```diff
@@ -0,0 +1,195 @@
+package arcs.core.analysis
+
+import arcs.core.data.expression.Expression
+
+/** Field [Identifier]. Lists of [Identifier]s imply an AccessPath.*/
+typealias Identifier = String
+
+/**
+ * Result of the [ExpressionClaimDeducer] on Paxel [Expression]s.
+ *
+ * [Deduction]s recursively associate identifiers to field paths. These relationships can be
+ * directly translated into [Claim]s and [AccessPath]s.
+ */
+sealed class Deduction {
+  /** Returns true if the [Deduction] collection has no members. */
+  open fun isEmpty(): Boolean = true
+
+  /** Returns true if the [Deduction] collection has some members. */
+  fun isNotEmpty(): Boolean = !isEmpty()
+
+  /** Unwrap [Deduction] object to get an underlying [Path]. */
+  abstract fun getPath(): Path
+
+  /** Substitute aliases in all paths. */
+  abstract fun substitute(aliases: Scope): Deduction
+
+  /** Union of two [Deduction] collections. */
+  abstract operator fun plus(other: Deduction): Deduction
+
+  /** [Deduction]s that have a simple path (a list of [Identifier]s). */
+  interface Pathlike {
+    val path: List<Identifier>
+    fun mergeTop(other: Pathlike): Deduction
+  }
+
+  /** A representation of a field path in a Paxel [Expression]. */
+  data class Path(override val path: List<Identifier> = emptyList()) : Deduction(), Pathlike {
```

Comment:
QQ: should `Path` and `Paths` be instances of a `Deduction` now? When (if ever) would you expect `Path` or `Paths` to be returned from the visitor directly?

Reply:
I'll think about it more, but here's one corner case: when we have a literal paxel expression.

Reply:
Here's an example:  246f60b